### ISSUE TITLE
MultiShotMemorySampler doesn't support multi headed inputs.

### ISSUE THREAD

**Issue by owenvallis** [tensorflow/similarity#324]
Currently we call tf.convert_to_tensor(np.array(x)) which assumes that x is a single headed input. We need to provide support for multiple inputs.

**Comment by owenvallis** [comment]
@dnazila, I looked in to [from_tensor_slice](https://github.com/tensorflow/tensorflow/blob/v2.12.0/tensorflow/python/data/ops/dataset_ops.py#L749) and I think this might be a solution that is also consistent with the standard TF Dataset API. 

Essentially, it looks like Keras expects a tuple of (X, y, sample_weights) for the fit and predict methods. The y and sample_weights components are optional. Additionally, you can nest tuples inside the main tuple to provide something like ((X1, X2, X3), (y1, y2), (w1, w2)) and it seems like it also supports dictionaries as well ({'a': X1, 'b': X2}, y).

The from_tensor_slice seems to also support the same Tuple structure as well.

So I think we could make the X argument of the sampler consistent with both Keras and the tf.dataset.from_tensor_slice. Additionally, the fit method states that 

`If x is a dataset, generator, or [keras.utils.Sequence](https://www.tensorflow.org/api_docs/python/tf/keras/utils/Sequence) instance, y should not be specified (since targets will be obtained from x).` 

So I think we can accept y if x is just a array-like or tensor and ignore y if x is tuple. This should make the change backwards compatible.

wdyt?

**Comment by dnazila** [comment]
Agreed, I think from_tensor_slice has a tf-consistent and well-specified behavior, so probably a good idea to defer to that

**Comment by owenvallis** [comment]
I think I have a solution using pure TF.data.Dataset objects. I need to run some speed tests as I've tried something similar in the past and it was much slower than our current in memory version. But I think this could work... and it would support the full set of input types and enable adding more TF.data.Dataset.map functions afterwards.

```
def key_func(x, y, *args):
  return y


def build_choices_ds(num_classes, num_examples_per_class):
  return tf.data.Dataset.random()\
    .map(lambda x: x%num_classes, name="mod_cid") \
    .map(lambda z: [[z]*num_examples_per_class], name="repeat_cid") \
    .flat_map(tf.data.Dataset.from_tensor_slices)


def build_similarity_dataset(tensor_slices, num_classes_per_batch, num_examples_per_class):
  raw_ds = tf.data.Dataset.from_tensor_slices(tensor_slices)

  window_size = len(raw_ds)
  group_by_cid = raw_ds.group_by_window(
            key_func,
            reduce_func=lambda key, ds: ds.batch(window_size),
            window_size=window_size,
  )

  cid_datasets = []
  for elem in group_by_cid.as_numpy_iterator():
    cid_datasets.append(tf.data.Dataset.from_tensor_slices(elem).repeat())
  choices_ds = build_choices_ds(len(cid_datasets), num_examples_per_class)

  batch_size = num_classes_per_batch * num_examples_per_class
  ds = tf.data.Dataset.choose_from_datasets(cid_datasets, choices_ds).repeat().batch(batch_size)
  return ds
  ```

**Comment by owenvallis** [comment]
Looks like it works! It's even a little faster as the Batch Size gets larger. This is a super simple net training on CIFAR100 using a batch size of 500 (10 classes * 50 examples per class):

![image](https://user-images.githubusercontent.com/916864/232274405-df31993a-e2ac-4183-b4a8-9b0d1cc0068d.png)

**Comment by owenvallis** [comment]
I also added a small shuffle
